### ISSUE TITLE
[Bug] FormatException: Input string was not in a correct format.

### ISSUE THREAD

**Issue by mgrogin** [firebase/quickstart-unity#1039]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2020.3.3f1
  * Firebase Unity SDK version: 7.1.0
  * Source you installed the SDK:.unitypackage 
  * Problematic Firebase Component: RemoteConfig
  * Other Firebase Components in use: Auth, Database, dynamic links, cloud message etc.)
  * Additional SDKs you are using: Facebook, ironsource, gamePlayService
  * Platform you are using the Unity editor on: Mac
  * Platform you are targeting: iOS
  * Scripting Runtime:IL2CPP)

### [REQUIRED] Please describe the issue here:

I'm seeing an exception in production for a lot of my users. I haven't been able to reproduce it
```
System.Number.StringToNumber (System.String str, System.Globalization.NumberStyles options, System.Number+NumberBuffer& number, System.Globalization.NumberFormatInfo info, System.Boolean parseDecimal) (at <00000000000000000000000000000000>:0)
System.Number.ParseInt64 (System.String value, System.Globalization.NumberStyles options, System.Globalization.NumberFormatInfo numfmt) (at <00000000000000000000000000000000>:0)
```

#### Steps to reproduce:
This has worked up until now. I changed all the functions to use FirebaseRemoteConfigDeprecated (I moved over from GPR to unityPackage)

`Firebase.RemoteConfig.FirebaseRemoteConfigDeprecated.GetValue(key).LongValue;` is throwing the exception

I'm only seeing this on IOS
on Android I saw this exception for a few users:
`
InitializationException: Firebase modules failed to initialize: auth (missing dependency), database (missing dependency), dynamic_links (missing dependency), messaging (missing dependency)`

**Comment by DellaBitta** [comment]
Hi @mgrogin,

Do you have information about the string value that's being converted to a long?  Could it be empty, or does it use any commas or periods to mark thousand places, etc?  Also, since this an error in the C# language libraries,  did you happen to change your .Net version from 3.5 to 4.0 by chance?


> on Android I saw this exception for a few users:
> InitializationException: Firebase modules failed to initialize: auth (missing dependency), database (missing dependency), dynamic_links (missing dependency), messaging (missing dependency) 

Could you post your initialization logic?  Something akin to [Start() in our quickstarts](https://github.com/firebase/quickstart-unity/blob/350c5db73fec8e849ac708f62e6604a7bdd5e35b/auth/testapp/Assets/Firebase/Sample/Auth/UIHandler.cs#L79)?

Thank you.

**Comment by mgrogin** [comment]
The Long values are short integer values like 5 and it worked before. I'm thinking maybe it's some how related to the fact that the SetDefaults function in FirebaseRemoteConfigDeprecated is async now and I was using it like before (sync) so maybe the setDefaults didn't finish?
          `  Firebase.RemoteConfig.FirebaseRemoteConfigDeprecated.SetDefaults(getRemoteConfigDefaultsDict());`
I don't think I was using .Net 3.5 since the previous versions of unity have been 2019.4

Right now while adjusting the SetDefaults function to work async I saw that I needed to Removed the Unity.Compat and Unity.Tasks dlls from the Parse/ root folder because of compile error: https://forum.unity.com/threads/type-task-exists-in-both-unity-tasks-and-mscorelib.855385/#post-5671555 but I don't know if that's related to any of this (Side note: In the GPR version of firebase I didnt need to worry about this)

As far as initialization logic All of the Firebase products (auth, realtime database, remoteconfig) we use call this function first (as you can see from the comment this code has been working for 3 years ):
```
    public void InitializeFirebaseAsyncWorkaround(Action action)
    {
        // WE have this function because theres an issue with calling  FirebaseApp.CheckDependencies() twice (currently and incorrectly tears down the all FirebaseApp objects) for android   https://github.com/firebase/quickstart-unity/issues/141 should be fixed within a week 19.03.2018
        try
        {
            app = Firebase.FirebaseApp.DefaultInstance;
            dependencyStatus = Firebase.DependencyStatus.Available;
            action();
        }
        catch (Firebase.InitializationException)
        {
            Firebase.FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task =>
            {
                dependencyStatus = task.Result;
                if (dependencyStatus == Firebase.DependencyStatus.Available)
                {
                    action();
                }
                else
                {
                    Debug.LogError(
                      "Could not resolve all Firebase dependencies: " + dependencyStatus);
                }
            });
        }
    }
```

Then I call something like this:
`InitializeFirebaseAsyncWorkaround(InitializeFirebaseRemoteConfig);`

**Comment by DellaBitta** [comment]
Hi @mgrogin,

> Right now while adjusting the SetDefaults function to work async I saw that I needed to Removed the Unity.Compat and Unity.Tasks dlls from the Parse/ root folder because of compile error: https://forum.unity.com/threads/type-task-exists-in-both-unity-tasks-and-mscorelib.855385/#post-5671555 but I don't know if that's related to any of this (Side note: In the GPR version of firebase I didnt need to worry about this)

The Parse issue is known and originates from our support of multiple .Net versions.  Our official workflow for fixing those errors can be found [Unity Setup Guide Troubleshooting section](https://firebase.google.com/docs/unity/setup#net_compatibility_when_using_unity_2017x_and_later).

> As far as initialization logic All of the Firebase products (auth, realtime database, remoteconfig) we use call this function first (as you can see from the comment this code has been working for 3 years ):
> 
> ...
> 

Could the Android logs that show an init error originate from the first call to `InitializeFirebaseAsyncWorkaround` as it attempts to use `Firebase.FirebaseApp.DefaultInstance` before calling the `CheckAndFixDependenciesAsync`? I'd suggest changing the detection of initialization to ensure that such a path doesn't exist -- for instance, create a flag in your app and initialize it to `false`.  Then in `InitializeFirebaseAsyncWorkaround` test the flag. If it's `false` call `CheckAndFixDependenciesAsync`, and set the flag to `true` in the task continuation.   If it's already `true` then use the `Firebase.FirebaseApp.DefaultInstance` since everything has already been initialized.

> The Long values are short integer values like 5 and it worked before. I'm thinking maybe it's some how related to the fact that the SetDefaults function in FirebaseRemoteConfigDeprecated is async now and I was using it like before (sync) so maybe the setDefaults didn't finish?

Yes, that seems like it could cause the strings to be empty and the the parse attempt could throw an exception in that case.  It might be worth logging the values before converting them, too, or log them in the catch of a try/catch block on a failure, just to ensure that they're valid Long values.
